File: vpn_ios/Shared/Preferences.swift
import Foundation

/// Когда приложение само поднимает туннель.
enum AutoConnectMode: String, CaseIterable, Codable {
    case off
    case always
    case cellularOnly
    case wifiOnly

    var title: String {
        switch self {
        case .off: return "Выключено"
        case .always: return "Всегда"
        case .cellularOnly: return "Только сотовая сеть"
        case .wifiOnly: return "Только Wi-Fi"
        }
    }
}

/// Настройки, общие для приложения, расширения и виджетов.
struct Preferences {
    private let defaults: UserDefaults

    init(defaults: UserDefaults) { self.defaults = defaults }

    static var shared: Preferences { Preferences(defaults: AppGroup.defaults ?? .standard) }

    var autoConnectMode: AutoConnectMode {
        get { AutoConnectMode(rawValue: defaults.string(forKey: Key.autoConnect) ?? "") ?? .off }
        nonmutating set { defaults.set(newValue.rawValue, forKey: Key.autoConnect) }
    }

    /// Сети, в которых туннель поднимать не нужно. Имена вводятся вручную:
    /// подставить текущее имя Wi-Fi мы могли бы только ценой доступа к геопозиции.
    var trustedNetworks: [String] {
        get { defaults.stringArray(forKey: Key.trusted) ?? [] }
        nonmutating set { defaults.set(newValue, forKey: Key.trusted) }
    }

    /// Последний известный баланс — чтобы виджет показывал его без похода в сеть.
    var lastBalance: String? {
        get { defaults.string(forKey: Key.balance) }
        nonmutating set { defaults.set(newValue, forKey: Key.balance) }
    }

    /// Включён ли фильтр рекламы и трекеров.
    var dnsFilter: Bool {
        get { defaults.bool(forKey: Key.dnsFilter) }
        nonmutating set { defaults.set(newValue, forKey: Key.dnsFilter) }
    }

    /// Настроен ли фильтр на сервере. Запоминается при получении конфигурации,
    /// чтобы экран настроек знал о доступности ещё до первого подключения
    /// после запуска приложения.
    var dnsFilterAvailable: Bool {
        get { defaults.bool(forKey: Key.dnsFilterAvailable) }
        nonmutating set { defaults.set(newValue, forKey: Key.dnsFilterAvailable) }
    }

    private enum Key {
        static let autoConnect = "autoConnectMode"
        static let trusted = "trustedNetworks"
        static let balance = "lastBalance"
        static let dnsFilter = "dnsFilter"
        static let dnsFilterAvailable = "dnsFilterAvailable"
    }
}

File: vpn_android/app/src/main/java/co/studio404/vpn/prefs/Preferences.kt
package co.studio404.vpn.prefs

import android.content.Context
import android.content.SharedPreferences
import org.json.JSONArray

enum class AutoConnectMode(val title: String) {
    OFF("Выключено"),
    ALWAYS("Всегда"),
    CELLULAR_ONLY("Только сотовая сеть"),
    WIFI_ONLY("Только Wi-Fi"),
}

class Preferences(context: Context) {
    private val prefs: SharedPreferences =
        context.getSharedPreferences("vpn_prefs", Context.MODE_PRIVATE)

    var autoConnectMode: AutoConnectMode
        get() = AutoConnectMode.entries
            .firstOrNull { it.name == prefs.getString(KEY_AUTO, null) }
            ?: AutoConnectMode.OFF
        set(value) = prefs.edit().putString(KEY_AUTO, value.name).apply()

    var trustedNetworks: List<String>
        get() {
            val raw = prefs.getString(KEY_TRUSTED, null) ?: return emptyList()
            return runCatching {
                val arr = JSONArray(raw)
                buildList {
                    for (i in 0 until arr.length()) {
                        val v = arr.optString(i).trim()
                        if (v.isNotEmpty()) add(v)
                    }
                }
            }.getOrDefault(emptyList())
        }
        set(value) {
            val arr = JSONArray()
            value.forEach { arr.put(it) }
            prefs.edit().putString(KEY_TRUSTED, arr.toString()).apply()
        }

    var dnsFilter: Boolean
        get() = prefs.getBoolean(KEY_DNS_FILTER, false)
        set(value) = prefs.edit().putBoolean(KEY_DNS_FILTER, value).apply()

    var dnsFilterAvailable: Boolean
        get() = prefs.getBoolean(KEY_DNS_AVAILABLE, false)
        set(value) = prefs.edit().putBoolean(KEY_DNS_AVAILABLE, value).apply()

    var lastBalance: String?
        get() = prefs.getString(KEY_BALANCE, null)
        set(value) = prefs.edit().putString(KEY_BALANCE, value).apply()

    companion object {
        private const val KEY_AUTO = "autoConnectMode"
        private const val KEY_TRUSTED = "trustedNetworks"
        private const val KEY_DNS_FILTER = "dnsFilter"
        private const val KEY_DNS_AVAILABLE = "dnsFilterAvailable"
        private const val KEY_BALANCE = "lastBalance"
    }
}
